From 3cda85f295d68981d0f769d3f7f352a3c9837491 Mon Sep 17 00:00:00 2001
From: Konstantin Porotchkin <kostap@marvell.com>
Date: Tue, 1 Sep 2020 20:22:56 +0300
Subject: [PATCH 0763/1921] arm64: dts: marvell: add SMMU support to CN913x
 family

Add IOMMU node for Marvell AP807 based SoCs (CN913x) along with
platform and PCI device Stream ID mapping.

Change-Id: I2da1aa072f53a08c02cf49a1528de7490c97a208
Signed-off-by: Konstantin Porotchkin <kostap@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/35144
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Stefan Chulski <stefanc@marvell.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 arch/arm64/boot/dts/marvell/cn9130-crb.dts |  5 +++++
 arch/arm64/boot/dts/marvell/cn9130-db.dts  |  5 +++++
 arch/arm64/boot/dts/marvell/cn9130.dtsi    | 21 +++++++++++++++++++++
 arch/arm64/boot/dts/marvell/cn9131-db.dts  |  2 ++
 4 files changed, 33 insertions(+)

diff --git a/arch/arm64/boot/dts/marvell/cn9130-crb.dts b/arch/arm64/boot/dts/marvell/cn9130-crb.dts
index 75201378f9d0..79fc88445b31 100644
--- a/arch/arm64/boot/dts/marvell/cn9130-crb.dts
+++ b/arch/arm64/boot/dts/marvell/cn9130-crb.dts
@@ -174,6 +174,11 @@
 		&cp0_comphy1 0
 		&cp0_comphy2 0
 		&cp0_comphy3 0>;
+	iommu-map =
+		<0x0   &smmu 0x480 0x20>,
+		<0x100 &smmu 0x4a0 0x20>,
+		<0x200 &smmu 0x4c0 0x20>;
+		iommu-map-mask = <0x031f>;
 };
 
 /* CON 28 */
diff --git a/arch/arm64/boot/dts/marvell/cn9130-db.dts b/arch/arm64/boot/dts/marvell/cn9130-db.dts
index ce49a70d88a0..bec942ce95d6 100644
--- a/arch/arm64/boot/dts/marvell/cn9130-db.dts
+++ b/arch/arm64/boot/dts/marvell/cn9130-db.dts
@@ -278,6 +278,11 @@
 		&cp0_comphy1 0
 		&cp0_comphy2 0
 		&cp0_comphy3 0>;
+	iommu-map =
+		<0x0   &smmu 0x480 0x20>,
+		<0x100 &smmu 0x4a0 0x20>,
+		<0x200 &smmu 0x4c0 0x20>;
+		iommu-map-mask = <0x031f>;
 };
 
 &cp0_sata0 {
diff --git a/arch/arm64/boot/dts/marvell/cn9130.dtsi b/arch/arm64/boot/dts/marvell/cn9130.dtsi
index a2b7e5ec979d..728d3fc04b5d 100644
--- a/arch/arm64/boot/dts/marvell/cn9130.dtsi
+++ b/arch/arm64/boot/dts/marvell/cn9130.dtsi
@@ -35,3 +35,24 @@
 #undef CP11X_PCIE0_BASE
 #undef CP11X_PCIE1_BASE
 #undef CP11X_PCIE2_BASE
+
+&smmu {
+	status = "okay";
+};
+
+&cp0_sata0 {
+	iommus = <&smmu 0x444>;
+};
+
+&cp0_sdhci0 {
+	iommus = <&smmu 0x445>;
+};
+
+&cp0_usb3_0 {
+	iommus = <&smmu 0x440>;
+};
+
+&cp0_usb3_1 {
+	iommus = <&smmu 0x441>;
+};
+
diff --git a/arch/arm64/boot/dts/marvell/cn9131-db.dts b/arch/arm64/boot/dts/marvell/cn9131-db.dts
index 3c975f98b2a3..f8582cec1ef2 100644
--- a/arch/arm64/boot/dts/marvell/cn9131-db.dts
+++ b/arch/arm64/boot/dts/marvell/cn9131-db.dts
@@ -123,6 +123,7 @@
 
 &cp1_sata0 {
 	status = "okay";
+	iommus = <&smmu 0x454>;
 
 	/* CON32 */
 	sata-port@1 {
@@ -195,6 +196,7 @@
 /* CON58 */
 &cp1_usb3_1 {
 	status = "okay";
+	iommus = <&smmu 0x451>;
 	usb-phy = <&cp1_usb3_0_phy0>;
 	/* Generic PHY, providing serdes lanes */
 	phys = <&cp1_comphy3 1>;
-- 
2.31.1

